metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala

```diff
@@ -360,25 +359,11 @@ class WorksheetProvider(
     val relativePath = path.toRelative(workspace)
     val evaluatedWorksheet =
       mdoc.evaluateWorksheet(relativePath.toString(), input.value)
-    val classpath = evaluatedWorksheet.classpath().asScala.toList
-    val previousDigest = worksheetsDigests.getOrElse(path, "")
-    val newDigest = calculateDigest(classpath)
     exportableEvaluations.update(
       input,
       evaluatedWorksheet,
     )
-
-    if (newDigest != previousDigest) {
-      worksheetsDigests.put(path, newDigest)
-      val sourceDeps = fetchDependencySources(
-        evaluatedWorksheet.dependencies().asScala.toSeq
-      )
-      compilers.restartWorksheetPresentationCompiler(
-        path,
-        classpath,
-        sourceDeps.filter(_.toString().endsWith("-sources.jar")),
-      )
-    }
+    maybeRestartWorksheetPresentationCompiler(path, evaluatedWorksheet)
```

Comment:
I wonder if instead of getting `evaluatedWorksheet` we could save the dependency sources together in the Compilers class? This would avoid the bidirectional flow between the two classes.

Reply:
I totally agree, the bidirectional flow is terrible here. So when we cancel and clear, we don't clean those saved dependencies? I was just a bit worried that it's a bit counterintuitive to leftover something after cancel. But it's still probably better then what we have now.

The other possibility I was considering was so that `WorksheetProvider` wouldn't create the `pc` but `Compilers` would create a new one for the worksheet if the digest went out of sync between `WorksheetProvider` and `Compilers`.

Reply:
> I was just a bit worried that it's a bit counterintuitive to leftover something after cancel. But it's still probably better then what we have now.

That's basically a target data (classpath etc.), which we also reuse after restarting, so that's fine.

> The other possibility I was considering was so that WorksheetProvider wouldn't create the pc but Compilers would create a new one for the worksheet if the digest went out of sync between WorksheetProvider and Compilers.

If we keep the digest in the Compilers, we should be fine? MAybe in something like `WorksheetData` that would contain both digest and source jars? And anything more needed to start the presentation compiler?

We would update it and maybe restart the compiler in case of change deps?